shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package co.nimblehq.avishek.phong.kmmic.presentation.module
+
+import co.nimblehq.avishek.phong.kmmic.data.remote.helper.extension.toErrorMessage
+import co.nimblehq.avishek.phong.kmmic.domain.usecase.LogInUseCase
+import co.nimblehq.avishek.phong.kmmic.helper.extension.isValidEmail
+import com.rickclephas.kmp.nativecoroutines.NativeCoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+
+data class LogInViewState(
+    val isSuccess: Boolean = false,
+    val isLoading: Boolean = false,
+    val error: String? = null,
+    val isInvalidEmail: Boolean = false,
+    var isInvalidPassword: Boolean = false
+) {
+    constructor() : this(false)
+
+    constructor(error: String?) : this(false, false, error)
+}
+
+class LogInViewModel(private val logInUseCase: LogInUseCase): BaseViewModel() {
+
+    private val mutableViewState: MutableStateFlow<LogInViewState> =
+        MutableStateFlow(LogInViewState())
+
+    val viewState: StateFlow<LogInViewState> = mutableViewState
+
+    fun logIn(email: String, password: String) {
+        if (!validInput(email, password)) { return }
+        setLoadingState()
+        viewModelScope.launch {
+            logInUseCase(email = email, password = password)
+                .catch { error -> handleLoginError(error) }
+                .collect { loginSuccess() }
+        }
+    }
+
+    private fun validInput(email: String, password: String): Boolean {
+        val isInvalidEmail = !email.isValidEmail()
+        val isInvalidPassword = password.isEmpty()
+        if (isInvalidEmail or isInvalidPassword)  {
+            mutableViewState.update {
+                LogInViewState(
+                    isInvalidEmail = isInvalidEmail,
+                    isInvalidPassword = isInvalidPassword
+                )
+            }
+            return false
+        }
+        return true
+    }
+
+    private fun setLoadingState() {
+        mutableViewState.update {
+            LogInViewState(isLoading = true)
+        }
+    }
+
+    private fun handleLoginError(error: Throwable) {
+        mutableViewState.update {
+            LogInViewState(isLoading = false, error = error.toErrorMessage())
```

Comment:
I was wondering, what if we need to show custom error messages on the app side instead of the API-retuned or the default throwable messages?

Reply:
Should we define a generic message in this case?  Do you have any idea?

Reply:
@AVI5HEK  I got your idea. However, this is quite big; can we bring it to another story? On iOS, I'm not using a key-value XML-based file to manage the localized string; we're using a 3rd party library and strong type for localized string. (e.g. `R.string.localizable.authenticationFieldsEmail()`)